### DISCUSSION TITLE
auto-save-state.lua中的问题

### DISCUSSION THREAD

**Discussion by zhongfly** [dyphire/mpv-config#103] [General]
1. demuxer-hysteresis-secs属性的问题

https://github.com/dyphire/mpv-config/blob/7d1cecca568eb20da913381d68e7c991e959bb3f/scripts/auto-save-state.lua#L41

我不太清楚为什么要判断demuxer-hysteresis-secs是否存在，但还是先只分析这一行的判断条件
`demuxer-hysteresis-secs`默认值是0，你可能希望是默认值0的时候，if会得到false的结果。然而实际上它在默认值时，`demuxer_secs = mp.get_property("demuxer-hysteresis-secs")`还是得到true

`mp.get_property("demuxer-hysteresis-secs")`得到的是字符串0.0000，需要用`mp.get_property_native("demuxer-hysteresis-secs")`，并且用`demuxer_secs~=0`来作判断

2. 解决完上面的问题，我猜测你的意图是设置了demuxer-hysteresis-secs为0时，播放结束时，执行`mp.commandv("delete-watch-later-config", path)`，即不保存watch-later。
实际上：先触发eof_reached函数，删除watch-later；然后播放器暂停pause，触发save_if_pause函数，又保存了watch-later。
这相当于watch-later还是保存了，eof_reached函数是无效的

我建议增加一个can_save变量

**Comment by dyphire** [comment]
> 1. demuxer-hysteresis-secs属性的问题
> 
> https://github.com/dyphire/mpv-config/blob/7d1cecca568eb20da913381d68e7c991e959bb3f/scripts/auto-save-state.lua#L41
> 
> 我不太清楚为什么要判断demuxer-hysteresis-secs是否存在，但还是先只分析这一行的判断条件 `demuxer-hysteresis-secs`默认值是0，你可能希望是默认值0的时候，if会得到false的结果。然而实际上它在默认值时，`demuxer_secs = mp.get_property("demuxer-hysteresis-secs")`还是得到true

我都记不得为什么添加 demuxer-hysteresis-secs 了，看了下 commit 历史才想起来。当时 mpv 对 watch-later-options 进行了破坏性的更新，然后脑抽了一下用临近添加的新属性 demuxer-hysteresis-secs 是否存在来进行兼容性判断了，实际应该只要判断 watch-later-options  参数就行了，在此更新之前 start 值无效

> 
> 2. 解决完上面的问题，我猜测你的意图是设置了demuxer-hysteresis-secs为0时，播放结束时，执行`mp.commandv("delete-watch-later-config", path)`，即不保存watch-later。
>    实际上：先触发eof_reached函数，删除watch-later；然后播放器暂停pause，触发save_if_pause函数，又保存了watch-later。
>    这相当于watch-later还是保存了，eof_reached函数是无效的
> 
> 我建议增加一个can_save变量

“播放器暂停 pause，触发 save_if_pause 函数”，这个应该只在 keep-open 不为 no 和 keep-open-pause=yes 时才会触发？我不使用 keep-open 而是 idle 所以没注意到这点

**Comment by zhongfly** [comment]
还有一个地方，我没搞清楚：
mp.observe_property("eof-reached", "bool", eof_reached)
mp.register_event("end-file", end_file)

为什么文件结束时需要监控2个属性/事件，感觉这2个应该会都触发？但实际上只见到前者的log

**Reply by dyphire** [reply]
eof-reached 只有 keep-open 不为 no 时才会触发，其实监控 end-file 应该就够了，有两个事件监控是原脚本的遗留

**Reply by zhongfly** [reply]
话说定时保存有啥意义吗？只有断电才需要吧

**Reply by dyphire** [reply]
应该只有意外退出时有用

**Reply by zhongfly** [reply]
我打算自己重写一个，去掉定时，只监控end-file。根据end-file的原因：如果是eof，删除进度；如果是其他原因(除了错误)，就保存进度…

**Reply by dyphire** [reply]
强杀进程的话会触发 end-file 吗？关机应该会

**Comment by zhongfly** [comment]
花时间测试了一下eof-reached属性、end-file事件，还有on_unload的hook

eof-reached属性用于判断文件播放到底是最及时、管用的，文件只要播放到底就立刻变true，加载文件前变nil，加载完文件变成false

end-file事件：在切换文件过程中，卸载文件时才触发这个事件，文件播放到底时还不会触发。
当keep-open=no|yes时，播放到底会继续播下一个文件，reason是eof
keep-open=always时，播放到底会暂停，手动跳到下一个文件或者是选择其他文件打开后触发end-file事件，reason就意想不到了，居然是stop
需要注意的是end-file事件时path属性已经是变成nil了

另外关闭播放器时，end-file的reason是quit，然后是shutdown事件。

on_unload的hook在end-file前，此时path属性还没清除，适合执行操作

**Reply by dyphire** [reply]
看上去 on_unload 更适合操作，eof-reached 受控于 keep-open 不是很好

**Reply by zhongfly** [reply]
on_unload的问题在于它是hook，我看文档，是没办法取消hook的…可能需要通过一个全局变量来间接开关

**Reply by dyphire** [reply]
使用 on_unload 的最大问题是无法区分应该执行删除还是保存操作，只使用 end-file 也一样，因为触发  end-file 事件时 path 属性已经是变成 nil 了，所以无法通过它实现文件 stop 或意外退出（杀死进程）时自动保存 watch-later 的功能

目前依然只能通过定时保存的方式来避免文件 stop 或意外退出（杀死进程）时未能保存 watch-later 的问题

**Comment by zhongfly** [comment]
完全重写了一个，去掉了定时，不过再加上去也很简单。
文件关闭时：
1. 进度是0或者全播完的话，删除进度，就算设置了save-position-on-quit也会被删除
2. 其他情况时，只有设置了save-position-on-quit才会保存进度

```lua
local msg = require("mp.msg")

local path = nil -- only set after file success load, reset to nil when file unload.

local function reset()
    path = nil	
end

-- set vars when file success load
local function init()
    path = mp.get_property("path")
end

-- save watch-later-config when file unloading
local function save_or_delete()
    local eof = mp.get_property_bool("eof-reached")
    local percent_pos = mp.get_property_number("percent-pos")
    if path ~= nil then
        if eof or percent_pos==0 then
            msg.debug("deleting state: percent_pos=0 or eof")
            mp.commandv("delete-watch-later-config", path)
        elseif mp.get_property_bool("save-position-on-quit") then
            msg.debug("saving state")
            mp.commandv("write-watch-later-config")
        end
    
        reset()
    end
end

mp.register_event("file-loaded", init)
mp.add_hook("on_unload", 50, save_or_delete) -- after mpv saving state
```

我自测感觉没啥问题，可以作为基础框架往上加脚本全局开关、当前文件临时关闭的功能

**Reply by dyphire** [reply]
这种方式无法实现对文件执行 stop 命令或意外退出（杀死进程）时自动保存 watch-later 的功能，因为不会触发 save-position-on-quit 的判定

**Reply by zhongfly** [reply]
> 这种方式无法实现对文件执行 stop 命令或意外退出（杀死进程）时自动保存 watch-later 的功能，因为不会触发 save-position-on-quit 的判定

保存进度是不依赖播放器自己的save-position-on-quit
我刚试了，stop可以保存进度
只要不是强行关闭程序，on_unload是文件关闭的必经之路

**Reply by zhongfly** [reply]
> 意外退出（杀死进程）时自动保存 watch-later

这种直接“断电”的方法只有定时保存能保证，别的都无法保证。

这个框架基础上加定时也很简单的，也可以不加在这个脚本里，因为没有耦合：定时保存发生在播放进行时，这个脚本的保存发生在播放结束时

**Reply by dyphire** [reply]
那么杀死进程的情况呢，不处理？

**Reply by zhongfly** [reply]
> 那么杀死进程的情况呢，不处理？

那种杀死进程？
具体的事件发生情况，你看看结果吧：
```
local utils = require("mp.utils")

local hooks = {"on_unload"}

for _, name in ipairs(hooks) do
    mp.add_hook(name, 0, function()
        print("--- hook: " .. name)
    end)
end

local events = {"file-loaded", "end-file","shutdown"}
for _, name in ipairs(events) do
    mp.register_event(name, function()
        print("--- event: " .. name )
    end)
end
```

如果这个脚本的没有成功保存，除了定时保存，你有其他方法吗？

**Reply by dyphire** [reply]
我刚才没有仔细看这个脚本，现在我要说它确实比 auto-save-state 脚本原有的相关实现好的多，不过我会保留定时保存功能

**Reply by zhongfly** [reply]
我的想法是，意外断电、程序崩溃这种特殊情况，交给定时保存来负责。这个脚本负责其他的，例如每5分钟定时保存，假如看了6分钟就切换到其他文件，这个6分钟的进度save-position-on-quit是不会保存的，但可以让这个脚本负责

**Reply by dyphire** [reply]
在提交中 https://github.com/dyphire/mpv-config/commit/7f83a8b 改用你提供的实现方式，并进行了调整

**Comment by zhongfly** [comment]
还有一个小细节，可能是keep-open=always或者最后一个文件才会有的问题：播放完毕，自动暂停时， percent-pos可能不是100，而是99.xx的一个接近100的数。同理，暂停时的时间也不会正好等于文件的时长，而是无限接近的值…

脚本的设置里percent_pos设为100可能会遇到这样的特殊情况不起作用，不过eof属性是足够作为判据，不需要再用percent_pos=100了

**Reply by dyphire** [reply]
将默认值改为 99
